### ISSUE TITLE
How to reproduce YOLOv8 evaluation results (mAP)?

### ISSUE THREAD

**Issue by rafi-fauzan** [ultralytics/ultralytics#3348]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, I trained the YOLOv8l model on my dataset, and these are the evaluation results on my test set:

							         map50
        all       7653     116714      0.791      0.705      0.687      0.528
       obj1       7653        898      0.692       0.99      0.964      0.853
       obj2       7653        159          1          0     0.0203     0.0199
       obj3       7653      17406      0.996      0.997      0.994      0.925
       obj4       7653        802      0.745      0.904      0.865      0.684
       obj5       7653       4854      0.969      0.995      0.986      0.847
       obj6       7653       1916      0.933      0.971      0.955      0.613
       obj7       7653       4727      0.418       0.24       0.24     0.0983
       obj8       7653       1719      0.435       0.36      0.279      0.103
       obj9       7653       2337      0.654      0.907       0.78      0.445
      obj10       7653       7653      0.996      0.999      0.995       0.99
      obj11       7653        425      0.471      0.727      0.597      0.289
      obj12       7653      31064      0.435      0.349      0.345       0.12
      obj13       7653       4974      0.814      0.931      0.931      0.632
      obj14       7653       4413      0.998      0.996      0.994      0.977
      obj15       7653         54          1          0    0.00505    0.00391
      obj16       7653       4400      0.981      0.994       0.99      0.843
      obj17       7653      28357      0.986      0.994      0.993      0.834
      obj18       7653        556      0.721       0.34      0.437      0.226

I attempted to reproduce the results using my trained model on the same test set by using the torchmetrics library (using default arguments for torchmetrics map, with iou_thresholds=[0.5]) but got different results, any idea why this might have occurred?

```
torchmetrics map50
{'obj1': 0.9576780796051025,
 'obj2': 0.0,
 'obj3': 0.9891015291213989,
 'obj4': 0.8018046617507935,
 'obj5': 0.9818739295005798,
 'obj6': 0.9419328570365906,
 'obj7': 0.1379074901342392,
 'obj8': 0.1875734031200409,
 'obj9': 0.7281174063682556,
 'obj10': 0.9896576404571533,
 'obj11': 0.5150798559188843,
 'obj12': 0.21758544445037842,
 'obj13': 0.8919346332550049,
 'obj14': 0.9892665147781372,
 'obj15': 0.0,
 'obj16': 0.9852380156517029,
 'obj17': 0.9885100722312927,
 'obj18': 0.26210492849349976}

```

### Additional

torchmetrics map documentation: https://torchmetrics.readthedocs.io/en/stable/detection/mean_average_precision.html

**Comment by glenn-jocher** [comment]
@rafi-fauzan dear user,

Thank you for reaching out. It seems that you are experiencing different results when trying to reproduce the evaluation metrics of your YOLOv8l model using the `torchmetrics` library. While `torchmetrics` is a useful library, it may not provide the exact same results as the evaluation metrics provided by the YOLOv8l model.

The evaluation results you obtained using YOLOv8l are based on a specific evaluation protocol and may involve additional post-processing steps specific to the YOLOv8 implementation. These steps go beyond the simple calculation of mean average precision (mAP) at an IoU threshold of 0.5.

When using `torchmetrics`, it is important to note that the default arguments for the `map` function may differ from the evaluation protocol used in YOLOv8. Additionally, the `torchmetrics` library might use different calculations, such as considering different IoU thresholds, that can affect the overall mAP score.

To accurately reproduce the evaluation results of your YOLOv8l model, you should consult the documentation or code specific to the YOLOv8 implementation you are using. The implementation details and evaluation protocol may differ between implementations, making it challenging to directly compare the results obtained using different libraries or tools.

If you have any further questions or concerns, please feel free to ask.

**Comment by vemundvestreng** [comment]
Hi! (short version below) 

I have a similar reproduction problem when trying to reproduce the validation results reported at the end of training with the "yolo detect val"-method. Although the returned metrics are comparable, they deviate beyond what I would assume to be rounding / numerical errors (e.g., recall on all classes of 0.508 reported at the end of training and 0.529 reported when running subsequent validation with the best weights and same dataset / mAP50 on all classes of 0.543 reported at the end of training and 0.553 reported when running subsequent validation with the best weights and same dataset). Do you have any idea why this might be? I can't see from the source code why they should deviate (when using the same "best.pt" weights and default validation settings), and this deviation does not appear to exist in between identical validation runs. I have, however, found that the validation mAP50 appears to vary with the IOU passed as an argument to the validation task, which I don't understand as I'd expect this to be constant? I assume (without being able to identify this directly in the source code) that the reason for this may be that the predictions below the passed IOU are discarded before the mAP50 is calculated, such that the reported mAP50 in truth is the mAP of the IOU passed as an argument (when this is above 0.5)?

As you can tell, I am struggling to definitely identify the confidence thresholds and IOUs used for the various metrics, which might be the cause for my confusion. As far as I can tell (please correct me if I'm wrong), the confusion matrix employs constant values of 0.25 and 0.45 for the confidence threshold and IOU, respectively (independent of the values passed as arguments to the validation). In ap_per_class, the returned precision and recall values appear use the confidence threshold that produces the maximum mean F1-score? The IOU used for these values appears to be 0.5 (0th row of the recall / precision matrices), but since the "tp" passed as input appears to already have been postprocessed with the non_max_suppression function using the confidence threshold and IOU passed as input arguments, I'm confused as to whether or not this means that the 0th row will equal the IOU passed as an input argument if this is above 0.5?

Short version:

1) Why does the validation metrics returned by the val method deviate from the ones returned at the end of training?
2) Why does the validation mAP50 apparently vary with the IOU passed as input?
3) What confidence threshold and IOU are used for the confusion matrix?
4) What confidence threshold and IOU are used for the precision and recall values?

I'm sorry to waste your time if I'm missing something obvious here, but I've been back-and-forth through the source code for some time now and am struggling to definitely find the answer to these questions - just want to make sure I'm evaluating / interpreting my results accurately!

**Comment by glenn-jocher** [comment]
@vemundvestreng hi!

Thank you for reaching out. I understand that you're experiencing some discrepancies between the validation results reported at the end of training with the "yolo detect val" method and the subsequent validation results using the best weights. I'll do my best to address your questions:

1) The deviation between the validation metrics reported at the end of training and the subsequent validation results could be due to a few factors. One possibility is that the validation dataset used at the end of training may differ slightly from the dataset used for subsequent validation. This could result in slight variations in the evaluation metrics. Another factor could be the randomness involved in the training process, which can lead to variations in the model's predictions.

2) The validation mAP50 may vary with the IOU passed as an input because the mAP calculation considers the detection results at different IOU thresholds. The mAP50 specifically calculates the average precision at an IOU threshold of 0.5. However, when you pass a different IOU threshold as an argument, the mAP calculation takes into account the precision and recall values at that specific IOU threshold.

3) The confidence threshold and IOU used for the confusion matrix are typically constant values set in the source code. In the YOLOv8 implementation, the confidence threshold is often set to 0.25 and the IOU threshold to 0.45. These values determine whether a prediction is considered a true positive or a false positive based on its confidence score and IOU with the ground truth.

4) The precision and recall values are calculated based on the confidence threshold that produces the maximum mean F1-score. The IOU used for these values is typically set to 0.5, which represents a relatively strict matching criterion. However, if you pass a different IOU threshold as an input, it will be considered when calculating precision and recall values.

It's important to note that the specific implementation details may vary, and different versions of YOLOv8 might have slight differences in the evaluation metrics. It's always a good idea to refer to the documentation or source code of the specific YOLOv8 implementation you are using for accurate information about the evaluation process and interpretation of results.

I hope this helps clarify your questions. If you have any further concerns or need additional assistance, please let me know.

**Comment by vemundvestreng** [comment]
Hi, @glenn-jocher! 

Thank you for a good and thorough answer. I have a couple of follow-ups on points 1 and 2:

1) How can the validation dataset used for validation during / after training vary from that used by the "yolo detect val" method? I am using the exact same dataset (i.e., identical images, identical labels, identical splits) with both methods. I understand that the batches and order in which they are processed may vary, but I don't understand how that would alter the final validation result? In terms of randomness, I thought that the "best.pt" weights contained a fixed set of parameters and that the network subsequently would be initialized without any random variables when passing "model='.../best.pt'"? I'd definitely expect the results to vary due to randomness in between different training runs on the same dataset, but not in between validations on the same dataset with the same set of weights (the results also appear to be consistent in between validation runs with the "yolo detect val" method, but there's a mis-match between those results and the ones reported at the end of training / in the results.csv-file).

2) Does that mean that if I pass iou=0.7 as an input argument to the training process, the validation mAP50 reported at the end of training / in the results.csv-file is really the mAP70 (and the mAP50-95 is really the mAP70-95) and the precision and recall values that are reported use 0.7 as an IOU threshold? I see that there are different default IOUs for the train and val methods (0.7 and 0.6, respectively), and so I thought maybe that could be the reason for the deviations in mAP50 experienced under point 1, but I still observe a deviation (also in precision and recall) when passing iou=0.7 to the val method.

**Comment by Rusab** [comment]
Is the IoU we are passing during training and validation the IoU used to determine True Positive? Or is it just used in the NMS process to reduce the prediction boxes?

**Comment by glenn-jocher** [comment]
@Rusab hi there! 

In YOLOv8, the IoU (Intersection over Union) parameter you pass during training and validation is primarily used in the non-maximum suppression (NMS) process. It helps determine which prediction boxes should be suppressed or discarded based on their overlap with other boxes. 

However, the IoU parameter does not directly affect the determination of True Positives. The True Positives are calculated based on the combination of the confidence threshold and the final predicted bounding boxes. The confidence threshold determines whether a prediction meets the required confidence score to be considered a positive detection. 

So, in summary, the IoU parameter you pass is mainly used in the NMS process to reduce redundant prediction boxes, while True Positives are determined by the confidence threshold and the final predicted bounding boxes. 

I hope this clarifies your question! If you have any further doubts, please let me know.

**Comment by Rusab** [comment]
Hi @glenn-jocher , thank you for the clarification.

**Comment by vemundvestreng** [comment]
> Hi! (short version below)
> 
> I have a similar reproduction problem when trying to reproduce the validation results reported at the end of training with the "yolo detect val"-method. Although the returned metrics are comparable, they deviate beyond what I would assume to be rounding / numerical errors (e.g., recall on all classes of 0.508 reported at the end of training and 0.529 reported when running subsequent validation with the best weights and same dataset / mAP50 on all classes of 0.543 reported at the end of training and 0.553 reported when running subsequent validation with the best weights and same dataset). Do you have any idea why this might be? I can't see from the source code why they should deviate (when using the same "best.pt" weights and default validation settings), and this deviation does not appear to exist in between identical validation runs. I have, however, found that the validation mAP50 appears to vary with the IOU passed as an argument to the validation task, which I don't understand as I'd expect this to be constant? I assume (without being able to identify this directly in the source code) that the reason for this may be that the predictions below the passed IOU are discarded before the mAP50 is calculated, such that the reported mAP50 in truth is the mAP of the IOU passed as an argument (when this is above 0.5)?
> 
> As you can tell, I am struggling to definitely identify the confidence thresholds and IOUs used for the various metrics, which might be the cause for my confusion. As far as I can tell (please correct me if I'm wrong), the confusion matrix employs constant values of 0.25 and 0.45 for the confidence threshold and IOU, respectively (independent of the values passed as arguments to the validation). In ap_per_class, the returned precision and recall values appear to use the confidence threshold that produces the maximum mean F1-score? The IOU used for these values appears to be 0.5 (0th row of the recall / precision matrices), but it's unclear to me if the "tp" passed as input already have been filtered / processed using some specific IOU threshold that would alter this interpretation.
> 
> Short version:
> 
> 1. Why does the validation metrics returned by the val method deviate from the ones returned at the end of training?
> 2. Why does the validation mAP50 apparently vary with the IOU passed as input?
> 3. What confidence threshold and IOU are used for the confusion matrix?
> 4. What confidence threshold and IOU are used for the precision and recall values?
> 
> I'm sorry to waste your time if I'm missing something obvious here, but I've been back-and-forth through the source code for some time now and am struggling to definitely find the answer to these questions - just want to make sure I'm evaluating / interpreting my results accurately!

Just a final update on this before the issue is closed if anyone has a similar question in the future: I was able to significantly reduce the observed deviation in validation performance when ensuring all the relevant validation parameters are the same in training and validation mode (the validation parameters default to different values between the modes). Especially the use of half precision appears to impact the validation results (argument "half" defaults to False in training mode but True in validation mode). Setting "half" to False when evaluating in validation mode and ensuring consistency also in the other parameters allowed to more-or-less reproduce the validation performance reported from the training mode.

**Comment by glenn-jocher** [comment]
@vemundvestreng i'm glad to hear that you’ve made progress on addressing the validation performance discrepancies by ensuring that the validation parameters are consistent between the training and validation modes. It's indeed important to use the same set of parameters for validation that are used at the end of training to achieve comparable results.

Using half precision (`half`) can introduce small numerical differences that might affect the evaluation metrics slightly. Even though these differences are usually minor, they can sometimes lead to noticeable variations in metrics like mAP, precision, and recall.

Thank you for sharing your solution with the community. Your findings about the impact of using half precision and the importance of maintaining consistent validation parameters will be valuable for others facing similar issues.

If you encounter any more questions or if there's anything else I can assist you with, feel free to reach out. Your contributions and thorough investigation enrich our shared understanding and use of the YOLOv8 model. Happy detecting!
